smart-py/objkt_swap.py

```diff
@@ -17,135 +38,336 @@ def __init__(self, objkt, hdao, manager, metadata, curate):
+
+            self.amount = self.objkt_amount * sp.fst(
+                sp.ediv(
+                    self.data.swaps[params.swap_id].xtz_per_objkt,
+                    sp.mutez(1)
+                ).open_some()
+            )
+
+            sp.verify(
+                (
+                    params.objkt_amount == self.objkt_amount
+                ) &
+                (
+                    sp.amount == sp.utils.nat_to_mutez(self.amount)
+                ) &
+                (
+                    sp.amount > sp.tez(0)
+                )
+            )
+
             # calculate fees and royalties
-            self.fee = sp.fst(sp.ediv(sp.utils.nat_to_mutez(self.amount), sp.utils.nat_to_mutez(1)).open_some()) * (self.data.royalties[self.data.swaps[params.swap_id].objkt_id].royalties + 25) / 1000
+            self.fee = sp.fst(
+                sp.ediv(
+                    sp.utils.nat_to_mutez(self.amount),
+                    sp.utils.nat_to_mutez(1)).open_some()
+                ) * (
+                    self.data.royalties[self.data.swaps[params.swap_id].objkt_id].royalties + 25
+                ) / 1000
+
             self.royalties = self.data.royalties[self.data.swaps[params.swap_id].objkt_id].royalties * self.fee / (self.data.royalties[self.data.swaps[params.swap_id].objkt_id].royalties + 25)
-            
+
             # send royalties to NFT creator
-            sp.send(self.data.royalties[self.data.swaps[params.swap_id].objkt_id].issuer, sp.utils.nat_to_mutez(self.royalties))
-            
+            sp.send(
+                self.data.royalties[self.data.swaps[params.swap_id].objkt_id].issuer,
+                sp.utils.nat_to_mutez(self.royalties)
+            )
+
             # send management fees
-            sp.send(self.data.manager, sp.utils.nat_to_mutez(abs(self.fee - self.royalties)))
-            
+            sp.send(
+                self.data.manager,
+                sp.utils.nat_to_mutez(
+                    abs(self.fee - self.royalties)
+                )
+            )
+
             # send value to issuer
-            sp.send(self.data.swaps[params.swap_id].issuer, sp.amount - sp.utils.nat_to_mutez(self.fee))
-            
+            sp.send(
+                self.data.swaps[params.swap_id].issuer,
+                sp.amount - sp.utils.nat_to_mutez(self.fee)
+            )
+
             # off on test scenarios
             # sp.if (sp.now < self.data.genesis):
             #self.mint_hDAO([sp.record(to_=sp.sender, amount=self.amount / 2), sp.record(to_=self.data.swaps[params.swap_id].issuer, amount=self.amount / 2), sp.record(to_=self.data.manager, amount=abs(self.fee - self.royalties))])
-        
-        self.fa2_transfer(self.data.objkt, sp.to_address(sp.self), sp.sender, self.data.swaps[params.swap_id].objkt_id, params.objkt_amount)
 
-        self.data.swaps[params.swap_id].objkt_amount = abs(self.data.swaps[params.swap_id].objkt_amount - params.objkt_amount)
-        
+        self.fa2_transfer(
```

Comment:
Any reentrant concerns since contract data are updated after external call?

Reply:
can you please elaborate i don't think i understand

Reply:
Add info to the earlier comment.